### ISSUE TITLE
ExtensionDtype API for preferred type when converting to NumPy array 

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#22791]
This is coming out of SparseArray. Right now if you have a homogeneous DataFrame of EAs, then doing something like `np.asarray(df)` will always convert to object. https://github.com/pandas-dev/pandas/blob/32a74f19b655f4b0ef198d5ff4e15d409c3670d2/pandas/core/internals/managers.py#L787-L794.

Should we give ExtensionDtype some input on the dtype used here? This would allow for some consistency between `np.asarray(series)` and `np.asarray(homogeneous_dataframe)`.

**Comment by jorisvandenbossche** [comment]
Not sure if that was from the same discussion, but I recall a discussion where we said that we basically wanted to know the "numpy dtype if I would be converted to a numpy array" of an ExtensionArray, without actually converting it. 
That seems something useful in general I think, and could then be used to determine the best interleaved dtype in the use case above?

**Comment by jorisvandenbossche** [comment]
It was indeed from the sparse discussion, my comment there: https://github.com/pandas-dev/pandas/pull/22325#discussion_r209979714

Basically, we want to know `np.array(EA).dtype` without doing the conversion to array. 

We could add a `numpy_dtype` attribute to the Dtype object?

This might also overlap with https://github.com/pandas-dev/pandas/issues/22224

**Comment by TomAugspurger** [comment]
There's some overlap with #22224, but I don't know if we can re-use the same attribute for both unfortunately. IntegerArray.numpy_type can't hold values with NaNs.

**Comment by jorisvandenbossche** [comment]
Yes, I don't think we can re-use it, but the `numpy_dtype` attribute of IntegerArray can still be changed of course.

The other issue is more about if you have a composite array, what is the dtype of the underlying array (which of course can also be multiple arrays .., so then this might also be multiple dtypes), while here it is the dtype when converted to a numpy array. 

But I would use a `numpy_dtype` name rather for the use case of this issue.

**Comment by jorisvandenbossche** [comment]
Although the question is, what if this dtype depends on the data? 
For example, for IntegerArray, we now decided to return an object array. But, you could also opt for actually returning the integer array if there are no NAs, and float otherwise.

**Comment by jreback** [comment]
I think we have this now? e.g. can just specify the ``dtype=`` in ``np.asarray()``, or something else needed here?

**Comment by jorisvandenbossche** [comment]
I don't think this is solved. It is not about passing a dtype in `np.asarray(..)`, the original question is: predicting for an array what the dtype will be of `np.array(dtype)` (so the default when not passing a `dtype=` kwarg).

One use case is to define the interleaved dtype (see top post)

(but it's not necessarily a blocker for 0.24.0)

**Comment by TomAugspurger** [comment]
Agreed: not solved, but a blocker.

**Comment by jbrockmendel** [comment]
interleaved_dtype goes through find_common_type, for which EA authors can specify behavior via _get_common_dtype.

There are some standard cases where find_common_type isn't quite right due to values-dependent-behavior.  Recently we've been collecting the values-dependent-behavior in dtypes.cast functions `find_result_type` and `common_dtype_categorical_compat`.  The latter could/should probably be extended to handle IntegerArray/BooleanArray with NAs.  I'd like to whittle these down into just `find_result_type`, but no immediate plans to do so.

So I'm thinking we could implement something like `EA._get_result_dtype_values_dependent`:`find_result_type`::`_get_common_type`:`find_common_type`

**Comment by JBGreisman** [comment]
This proposal sounds related to #22224 -- would the proposed plan be to use something like `EA.find_result_type` here: 
https://github.com/pandas-dev/pandas/blob/60ac973e64ba364b345aea22bf988272536aa42f/pandas/core/internals/managers.py#L1587-L1595

in the clause for `ExtensionDtype`?

**Comment by MarcoGorelli** [comment]
> _get_result_dtype_values_dependent

is this really desirable though? wouldn't it go against [this comment](https://github.com/pandas-dev/pandas/issues/30038#issuecomment-613427298):

> We try to avoid value-dependent behavior where the metadata (shape, dtype, etc.) depend on the values of the array.

I'd be more inclined to do:
1. find the interleaved type. E.g. if `Int64Dtype()` and `Int32Dtype()`, then this would be `Int64Dtype()`
2. convert to the corresponding `numpy` type (`int64`)
3. if there are missing values (which can't be represented by `int64`), then raise. It's then up to the user to provide a suitable `dtype` (e.g. `float64`) and/or a suitable `na_value`

This would be the 2D analogue of https://github.com/pandas-dev/pandas/issues/48891
